File: licakim/week3/selection_sort.c
/*
선택 정렬 : 현재 레코드 뒤의 레코드 중 가장 작은 인덱스만 기억하여 현재 레코드와 교환하며 정렬
공간 복잡도:  θ(1)
-> 입력배열의 크기만이 사용되고 추가 저장장소 사용량이 없는 제자리 정렬
시간 복잡도: O(n^2)
-> 비교 횟수(내부루프) T(n) = (n-1)+(n-2)+...+1 = n(n-1)/2
   교환 횟수 (외부루프) T(n) = 3(n-1) n-1번 교환하며 3번의 저장(swap)이 필요 
*/

void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

void selection_sort(int n, int *arr)
{
    int i,j;
    int min;

    for(i = 0; i < n - 1; i++)
    {
        min = i;
        for(j = i + 1 ; j < n; j++)
        {
            if(arr[j] < arr[min])
            {
                min = j;
            }
        }
        swap(&arr[min],&arr[i]);
    }
}

File: Jimin0304/week3/SelectionSort.py
# 공간복잡도 : 주어진 배열이 차지하고 있는 공간 내에서 값들의 위치만 바꾸기 때문에 O(1)
# 시간복잡도 :
# 루프문을 통해 모든 인덱스에 접근해야 하기 때문에 기본적으로 O(N)을 시간을 소모,
# 하나의 루프에서는 현재 인덱스의 값과 다른 인덱스의 값들과 비교하여 최소값을 찾은 후 
# 현재 인덱스에 있는 값과 상호 자리 교대를(swap)해야 해야하기 때문에 O(N)을 시간이 필요
# 즉, O(N^2)의 시간 복잡도

def selection_sort(arr):
    num = len(arr)
    for i in range(0, num - 1):
        min_idx = i
        for j in range(i + 1, num):
            if arr[min_idx] > arr[j]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]

arr = [3, 5, 1, 4, 2]
print("original:", arr)
selection_sort(arr)
print("selection sort:", arr)